### ISSUE TITLE
Array gen disregard size

### ISSUE THREAD

**Issue by jutaro** [purescript/purescript-quickcheck#20]
As far as I see, the array generator works without care for size:

~~~
instance arbArray :: (Arbitrary a) => Arbitrary [a] where
  arbitrary = do
    b <- arbitrary
    if b then return [] else do
      a <- arbitrary
      as <- arbitrary
      return (a : as)
~~~

When I do look to the Haskell quickcheck lib things look well behaved to me:

~~~
instance Arbitrary a => Arbitrary [a] where
  arbitrary = sized $ \n ->
    do k <- choose (0,n)
       sequence [ arbitrary | _ <- [1..k] ]
~~~

Shall I provide a patch? (I guess that 'choose' in purescript will produce floats always).

**Comment by jdegoes** [comment]
:+1: I'd also merge this into [purescript-strongcheck](purescript-contrib/purescript-strongcheck)

**Comment by paf31** [comment]
What would the replacement be? We don't have a concept of size in purescirpt-quickcheck.

**Comment by jdegoes** [comment]
Yes, I added sizing, and it's one of the changes to survive the revert.

**Comment by paf31** [comment]
Oops :) I say go for it then :+1:

**Comment by jutaro** [comment]
In this function size is already used, and it is the same I would do in the Arbitrary instance:

~~~
arrayOf :: forall a. Gen a -> Gen [a]
arrayOf g = sized $ \n ->
  do k <- chooseInt 0 n
     vectorOf k g
~~~

As I understand it, in Haskell quickcheck test cases are generated with increased size. I can find it in sample, (but I don't know where it is done in 'real' quickcheck):

~~~
-- | Generates some example values.
sample' :: Gen a -> IO [a]
sample' g =
  generate (sequence [ resize n g | n <- [0,2..20] ])
~~~

However, this is maybe missing in purescript quickcheck, e.g. the sample
function always sets a fixed size:

~~~
-- | Samples a generator, producing the specified number of values. Uses 
-- | default settings for the initial generator state.
sample :: forall f a. (Monad f) => Number -> GenT f a -> f [a]
sample n = sample' n (GenState { size: 10, seed: 4545645874 })  
~~~

Am I right, that it would make sense to increase the complexity of generated test cases with the size parameter. How does strongcheck work in this regard?

**Comment by jdegoes** [comment]
stronghcheck doesn't yet do that, but it's on my list of todos before the 1.0 release comes out.

strongcheck also has other abilities, like finding more failures from a given one.

**Comment by sharkdp** [comment]
I think this can be closed, due to 392134e9bbc4de13f0bfee778929a1d4fdbcfc84 ?

**Comment by garyb** [comment]
Yep, I think so!
